pgjdbc/src/main/java/org/postgresql/core/v3/ConnectionFactoryImpl.java

```diff
@@ -116,21 +116,27 @@ public QueryExecutor openConnectionImpl(HostSpec[] hostSpecs, String user, Strin
     int connectTimeout = PGProperty.CONNECT_TIMEOUT.getInt(info) * 1000;
 
     // - the targetServerType setting
-    HostRequirement targetServerType;
-    String targetServerTypeStr = PGProperty.TARGET_SERVER_TYPE.get(info);
-    try {
-      targetServerType = HostRequirement.valueOf(targetServerTypeStr);
-    } catch (IllegalArgumentException ex) {
-      throw new PSQLException(
-          GT.tr("Invalid targetServerType value: {0}", targetServerTypeStr),
-          PSQLState.CONNECTION_UNABLE_TO_CONNECT);
+    HostRequirement requestedTargetServerType = getTargetServerType(info);
+
+    HostRequirement targetServerType = requestedTargetServerType;
+    HostRequirement nextTargetServerType = null;
+    if (targetServerType == HostRequirement.preferSlave) {
+      targetServerType = HostRequirement.slave;
+      nextTargetServerType = HostRequirement.any;
     }
 
     SocketFactory socketFactory = SocketFactoryFactory.getSocketFactory(info);
 
     HostChooser hostChooser =
         HostChooserFactory.createHostChooser(hostSpecs, targetServerType, info);
     Iterator<HostSpec> hostIter = hostChooser.iterator();
+
+    if (!hostIter.hasNext() && nextTargetServerType != null) {
```

Comment:
This code gets repeated multiple times and that is a bad sign as that somewhat duplicate the concept of `iterator()` that has been there.

Reply:
i had refactor to avoid repeated  those code, please review it again.